cranelift/codegen/src/isa/aarch64/lower.rs

```diff
@@ -202,12 +202,12 @@ pub(crate) fn output_to_reg<C: LowerCtx<I = Inst>>(ctx: &mut C, out: InsnOutput)
 /// The given register will be extended appropriately, according to
 /// `narrow_mode` and the input's type. If extended, the value is
 /// always extended to 64 bits, for simplicity.
-pub(crate) fn input_to_reg<C: LowerCtx<I = Inst>>(
+pub(crate) fn put_input_in_reg<C: LowerCtx<I = Inst>>(
```

Comment:
Thanks for this renaming which clarifies the side-effects happening here indeed.

As an alternative, as demonstrated by the changes in this PR, it seems that the narrowing is the edge case, and not the usual thing to do.

Two propositions:

- either we could add another helper function `get_input(ctx, input)` which really reads the input and does no side-effect? Callers of `put_input_in_reg` with `NarrowMode::None` could call this function instead.
- or we could just remove the NarrowMode parameter all the time, and having another helper function that does the narrowing, separating the different responsibilities more clearly? (I hope to implement it this way on the x64 backend, if possible.)

Reply:
Sure, I would be fine with the latter; perhaps we can consider this in the context of refactoring to share helpers with the x64 backend? (I'm thinking that e.g. we could define a trait to describe codegen for value widening and then plug that into a generic helper, for example. Similar for codegenning constants. Part of a "libMachBackend" of sorts as an optional layer on top of the lowering context API.)

Reply:
Good idea!